File: python/mail.py
import base64
import getopt
import imaplib
import email
import json
import random
import smtplib
import time
import sys
import re

from email.header import decode_header
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

OK = "OK"

OP_REPLY = "reply"
OP_DEL = "delete"
OP_READ = "read"
OP_MOVE = "move"

# account credentials
username = ""
password = ""

html = """
<div>
    <div>
        {reply}
    </div>
    <blockquote style="margin:0px 0px 0px 0.8ex;
        border-left:2px solid rgb({r},{g},{b});
        padding-left:1ex">
        {origin}
    </blockquote>
<div>
"""
plain = """
{reply}
{origin}
"""


def get_text_payload(msg: email.message.Message) -> dict:
    encode = msg.get("Content-Transfer-Encoding", None)
    payload = msg.get_payload()
    if encode and encode == "base64":
        return {
            "type": msg.get_content_subtype(),
            "data": base64.standard_b64decode(payload).decode('utf-8', errors='ignore'),
        }
    else:
        return {
            "type": msg.get_content_subtype(),
            "data": payload,
        }


def get_multipart_payload(msg: email.message.Message) -> dict:
    ans = list()
    for part in msg.get_payload():
        ans.append(get_text_payload(part))

    return {
        "type": "multipart",
        "data": ans
    }


def get_plain_body(msg: email.message.Message) -> str:
    body = ''
    maintype = msg.get_content_maintype()
    if maintype == "multipart":
        multi = get_multipart_payload(msg).get("data", [])
        for text in multi:
            if text.get("type") == "html":
                body = text.get("data", "")
            if text.get("type") == "plain" and body == "":
                body = text.get("data", "")
    elif maintype == "text":
        body = get_text_payload(msg).get("data", "")

    return body


def get_body(message: email.message.Message) -> dict:
    body = {}
    maintype = message.get_content_maintype()
    if maintype == "multipart":
        body = get_multipart_payload(message)
    elif maintype == "text":
        body = get_text_payload(message)

    return body


def get_header(message: email.message.Message, field: str) -> str:
    value = message.get(field, None)
    if not value:
        return ''
    try:
        tokens = decode_header(value)
        if len(tokens) == 0:
            return ''

        value = ''
        for token in tokens:
            subject = token[0]
            encode = token[1]
            if not encode or encode in ["unknown-8bit"]:
                encode = "utf-8"
            if isinstance(subject, bytes):
                value += " " + subject.decode(encoding=encode)
            elif isinstance(subject, str):
                value += " " + subject
    except Exception as err:
        print(err)

    return value.lstrip(" ")


def login() -> imaplib.IMAP4_SSL:
    # create an POP class with SSL
    ssl = imaplib.IMAP4_SSL("imap.qq.com", port=993)
    # authenticate
    status, data = ssl.login(username, password)
    if status != OK:
        print(data)
        exit(1)

    # select the mailbox I want to delete in
    # if you want SPAM, use imap.select("SPAM") instead
    status, data = ssl.select("INBOX")
    if status != OK:
        print(data)
        exit(1)

    return ssl


def close(imap: imaplib.IMAP4):
    # close the mailbox
    imap.close()
    # logout from the account
    imap.logout()


def tag(imap: imaplib.IMAP4_SSL, mail: bytes, flags: str):
    return imap.store(mail, "+FLAGS", flags)


def remove(imap: imaplib.IMAP4_SSL, mail: bytes):
    tag(imap, mail, "\\Deleted")
    return imap.expunge()


def move(imap: imaplib.IMAP4_SSL, mail: bytes, folder: str):
    result = imap.copy(mail, folder)
    if result[0] == OK:
        return remove(imap, mail)
    return result


def readed(imap: imaplib.IMAP4_SSL, mail: bytes):
    return tag(imap, mail, "\\Seen")


def default_reply_message(mail: email.message.Message) -> str:
    sender = get_header(mail, "From")
    match = re.search(r'<(.*?)@.*?>', sender, re.IGNORECASE)
    if match:
        return "Hello, " + match.groups()[0]
    else:
        return "Hello, " + sender


def reply(mail: email.message.Message, config: dict):
    new = MIMEMultipart("alternative")
    new["Date"] = time.strftime("%a, %d %b %Y %H:%M:%S +0800 (CST)")
    new["Message-ID"] = email.utils.make_msgid()
    new["In-Reply-To"] = mail["Message-ID"]
    new["Subject"] = "Re: " + mail["Subject"]
    new["To"] = mail["Reply-To"] or mail["From"]
    new["From"] = mail["To"]

    reply_message = default_reply_message(mail)
    plainbody = config.get("body", reply_message)
    htmlbody = config.get("body", "<div>" + reply_message + "</div>")

    origin = get_body(mail)
    originplain = ""
    originhtml = ""
    maintype = origin.get("type")
    if maintype == "plain":
        originplain += origin.get("data") + "\n"
    elif maintype == "html":
        originhtml += origin.get("data") + "<br>"
    elif maintype == "multipart":
        data = origin.get("data", [{}, {}])
        for idx in data:
            typ = idx.get("type")
            maintype = (lambda x: x if maintype == "multipart" else maintype)(typ)
            if typ == "plain":
                originplain += idx.get("data", "") + "\n"
            elif typ == "html":
                originhtml += idx.get("data", "") + "<br>"
    else:
        maintype = "plain"

    if maintype == "html":
        originhtml = originhtml[:-len("<br>")]
        r, g, b = random.sample(range(1, 255), 3)
        body = html.format(reply=htmlbody, origin=originhtml, r=r, g=g, b=b)
        new.attach(MIMEText(body, "html", "UTF-8"))
    else:
        originplain = originplain[:-len("\n")]
        body = plain.format(reply=plainbody, origin=originplain)
        new.attach(MIMEText(body, "plain", "UTF-8"))

    print(json.dumps(get_body(new),
                     ensure_ascii=False,
                     sort_keys=False,
                     indent=3,
                     separators=(", ", ": ")))

    from html.parser import HTMLParser
    body = get_body(mail)
    if body.get("type") == "multipart":
        print(HTMLParser().unescape(body.get("data")[0]))
    elif body.get("type") == "text":
        print(HTMLParser().unescape(body.get("data")))

    s = smtplib.SMTP("smtp.qq.com", 587)
    result = s.login(username, password)
    print(result)
    result = s.sendmail(mail["To"], [new["To"]], new.as_string())
    print(result)
    s.quit()


def regex(rules):
    if isinstance(rules, list):
        ans = []
        for r in rules:
            ans.append(re.compile(r, flags=re.IGNORECASE | re.UNICODE | re.MULTILINE))
        return ans
    elif isinstance(rules, str):
        return re.compile(rules, flags=re.IGNORECASE | re.UNICODE | re.MULTILINE)
    else:
        return re.compile('', flags=re.IGNORECASE | re.UNICODE | re.MULTILINE)


def match(rules, data: str) -> bool:
    if isinstance(rules, list):
        for r in rules:
            if r.search(data):
                return True
        return False
    else:
        return rules.search(data) is not None


# search for attr(FROM/SUBJECT/TO/CC)
# status, messages = imap.search(None, "FROM <925184024@qq.com>")
# search for header
# status, messages = imap.search(None, "HEADER Message-ID "xxxx"")
# search for status
# status, messages = imap.search(None, "(UNSEEN)")
# seach ALL
# status, messages = imap.search(None, "ALL")
# search for date(BEFORE/AFTER)
# status, messages = imap.search(None, "BEFORE "01-JAN-2020"")
def handle(imap: imaplib.IMAP4_SSL, config: list):
    condition = list()
    starttime = '9999-12-31'
    endtime = '0000-00-00'
    for c in config:
        print("job config:", c.get("name"))
        now = time.strftime("%Y-%m-%d")
        start = time.strftime("%Y-%m-%d", time.strptime(c.get("since", now), "%Y-%m-%d"))
        end = time.strftime("%Y-%m-%d", time.strptime(c.get("before", now), "%Y-%m-%d"))
        print("date range: [%s -> %s]" % (start, end))
        starttime = min(starttime, start)
        endtime = max(endtime, end)
        condition.append({
            "start": start,
            "end": end,
            "rbody": regex(c.get("body")),
            "rfrom": regex(c.get("from")),
            "rsubj": regex(c.get("subject")),
            "type": c.get("op", {}).get("type"),
        })

    # mails
    start = time.strftime("%d-%b-%Y", time.strptime(starttime, "%Y-%m-%d"))
    end = time.strftime("%d-%b-%Y", time.strptime(endtime, "%Y-%m-%d"))
    status, messages = imap.search(None, 'SINCE "%s"' % (start,), 'BEFORE "%s"' % (end,))
    if status != OK:
        print("search failure", status)
        return

    messages = messages[0]
    if messages == b'':
        print("no emails")
        return

    # convert messages to a list of email IDs
    messages = messages.split(b" ")
    print("total mails: %d" % (len(messages)))

    for uid in messages:
        _, data = imap.fetch(uid, "(RFC822)")
        # you can delete the for loop for performance if you have a long list of emails
        # because it is only for printing the SUBJECT of target email to delete
        for response in data:
            if isinstance(response, tuple):
                #  tuple, 0: RFC 1: msg
                message = email.message_from_bytes(response[1])
                # decode the email subject

                body = get_plain_body(message)
                subject = get_header(message, "Subject")
                sender = get_header(message, "From")
                to = get_header(message, "To")
                date = get_header(message, "Date")

                for cond in condition:
                    # date
                    try:
                        start = cond.get("start")
                        end = cond.get("end")
                        if date.endswith(" (CST)"):
                            date = date[0:-6]
                        # Tue, 9 Mar 2021 19:10:03 +0800 (CST)
                        style = "%a, %d %b %Y %H:%M:%S %z"
                        day = time.strftime("%Y-%m-%d", time.strptime(date, style))
                        if day < start or day > end:
                            continue
                    except ValueError:
                        continue

                    # regex
                    rbody = cond.get("rbody")
                    rfrom = cond.get("rfrom")
                    rsubj = cond.get("rsubj")
                    op = cond.get("type")

                    if match(rfrom, sender) and match(rsubj, subject) and match(rbody, body):
                        if op == OP_REPLY:
                            print("[{to}] will send mail to [{send}]".format(to=to, send=sender))
                            reply(message, c.get("op", {}))
                        elif op == OP_DEL:
                            print("delete mail from [{send}]".format(send=sender))
                            remove(imap, uid)
                        elif op == OP_READ:
                            print("tag mail to READ from [{send}]".format(send=sender))
                            readed(imap, uid)
                        elif op == OP_MOVE:
                            print("move mails from [{send}]".format(send=sender))
                            move(imap, uid, c.get("folder", "INBOX"))


if __name__ == "__main__":
    argv = sys.argv[1:]

    try:
        opts, args = getopt.getopt(argv, "u:p:", ["username=", "password="])  # 长选项模式
        for opt, arg in opts:
            if opt in ["-u", "--username"]:
                username = arg
            if opt in ["-p", "--password"]:
                password = arg
    except Exception as e:
        print("param error", e)

    if not username or not password:
        print("no username or password")
        exit(1)

    client = login()
    handle(client, list([{
        "name": "github",
        "since": "2021-01-10",
        "before": "2022-04-18",
        "from": ["github.com"],
        "body": ["jobs have failed", "refs/heads",
                 "No jobs were run", "Successfully deployed",
                 "Deployment failed", "A personal access token"],
        "op": {
            "type": OP_DEL
        }
    }, {
        "name": "google",
        "since": "2021-01-10",
        "before": "2022-04-18",
        "from": ["no-reply@accounts.google.com"],
        "subject": ["安全提醒", "帐号的安全性"],
        "op": {
            "type": OP_DEL
        }
    }, {
        "name": "jd",
        "since": "2021-01-10",
        "before": "2022-04-18",
        "from": ["newsletter1@jd.com"],
        "op": {
            "type": OP_DEL
        }
    }]))
    close(client)

File: golang/mail.go
package main

import (
	"bufio"
	"bytes"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/emersion/go-imap/commands"
	"io"
	"io/ioutil"
	"mime/quotedprintable"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/emersion/go-imap"
	"github.com/emersion/go-imap/client"
	"golang.org/x/text/encoding/simplifiedchinese"
	"golang.org/x/text/transform"
)

type Email struct {
	Username string
	Password string
	client   *client.Client
}

type operate struct {
	Type    string `json:"type"`
	MailBox string `json:"mailbox"`
}

type config struct {
	Name    string    `json:"name"`
	Since   time.Time `json:"since"`
	Before  time.Time `json:"before"`
	From    []string  `json:"from"`
	Body    []string  `json:"body"`
	Subject []string  `json:"subject"`
	Operate operate   `json:"op"`
}

type condition struct {
	start   time.Time
	end     time.Time
	rbody   []*regexp.Regexp
	rfrom   []*regexp.Regexp
	rsubj   []*regexp.Regexp
	operate operate
}

const (
	OPDEL  = "delete"
	OPREAD = "read"
	OPMOVE = "move"
)

func regex(str []string) []*regexp.Regexp {
	var ans = make([]*regexp.Regexp, len(str))
	for i := range str {
		ans[i] = regexp.MustCompile(str[i])
	}
	if len(ans) == 0 {
		ans = append(ans, regexp.MustCompile(".*"))
	}

	return ans
}

func match(str string, r []*regexp.Regexp) bool {
	for _, v := range r {
		if v.MatchString(str) {
			return true
		}
	}

	return false
}

func (e *Email) Login() error {
	var err error
	e.client, err = client.DialTLS("imap.qq.com:993", &tls.Config{
		ServerName: "imap.qq.com",
	})
	if err != nil {
		return err
	}

	err = e.client.Login(e.Username, e.Password)
	if err != nil {
		return err
	}

	return nil
}

func (e *Email) Handle(configs []config) error {
	var conds []condition
	start := time.Now().Add(10 * 365 * 24 * time.Hour)
	end := time.Unix(0, 0)

	now := time.Now()
	for _, config := range configs {
		fmt.Println("job config:", config.Name)
		fmt.Printf("date range: [%s -> %s]\n", config.Since, config.Before)
		if config.Since.After(now) {
			config.Since = now
		}
		if config.Before.Before(now) {
			config.Before = now
		}

		if start.After(config.Since) {
			start = config.Since
		}
		if end.Before(config.Before) {
			end = config.Before
		}
		conds = append(conds, condition{
			start:   config.Since,
			end:     config.Before,
			rbody:   regex(config.Body),
			rfrom:   regex(config.From),
			rsubj:   regex(config.Subject),
			operate: config.Operate,
		})
	}

	if len(configs) == 0 {
		start, end = end, start
	}

	_, err := e.client.Select("INBOX", false)
	if err != nil {
		return err
	}

	search := &imap.SearchCriteria{
		Since:  start,
		Before: end,
	}
	nums, err := e.client.Search(search)
	if err != nil {
		return err
	}

	fmt.Println("mail len:", len(nums))
	seqset := new(imap.SeqSet)
	seqset.AddNum(nums...)
	ch := make(chan *imap.Message, 10)
	done := make(chan error, 1)
	go func() {
		done <- e.client.Fetch(seqset, []imap.FetchItem{
			imap.FetchEnvelope,
			imap.FetchRFC822Text,
		}, ch)
	}()

	for msg := range ch {
		message, err := e.ParseMessage(msg)
		if err != nil {
			continue
		}
		e.handleMessage(message, conds)
	}

	if err := <-done; err != nil {
		return err
	}

	return nil
}

const (
	HeaderType   = "Content-Type"
	HeaderCoding = "Content-Transfer-Encoding"
)

type Message struct {
	Date    time.Time
	From    string
	To      string
	Subject string
	Body    string
	Origin  *imap.Message
}

func (e *Email) ParseMessage(message *imap.Message) (msg Message, err error) {
	env := message.Envelope
	msg.Subject = env.Subject
	msg.Date = env.Date
	msg.To = env.To[0].MailboxName + "@" + env.To[0].HostName
	msg.From = env.From[0].MailboxName + "@" + env.From[0].HostName
	msg.Origin = message

	for _, val := range message.Body {
		var buf bytes.Buffer
		io.Copy(&buf, val)
		backup := buf.String()
		reader := bufio.NewReader(&buf)
		var (
			line, ctype, coding, data string
			begin                     bool
		)
		brline, err := reader.ReadString('\n')
		for err == nil && line != brline {
			if begin {
				data += line
			}
			if strings.HasPrefix(line, HeaderType) {
				ctype = strings.Split(line, ":")[1]
				ctype = strings.TrimSpace(ctype)
			}
			if strings.HasPrefix(line, HeaderCoding) {
				coding = strings.Split(line, ":")[1]
				coding = strings.TrimSpace(coding)
				begin = true
			}
			line, err = reader.ReadString('\n')
		}
		if err != nil && err != io.EOF {
			return msg, err
		}

		// base64, quoted-printable
		var raw []byte
		data = strings.TrimSpace(data)
		if coding == "base64" {
			raw, err = base64.StdEncoding.DecodeString(data)
		} else if coding == "quoted-printable" {
			var reader io.Reader
			reader = quotedprintable.NewReader(bytes.NewBufferString(data))
			if strings.Contains(ctype, "gbk") {
				reader = transform.NewReader(reader, simplifiedchinese.GBK.NewEncoder())
			}
			raw, err = ioutil.ReadAll(reader)
		} else {
			raw, err = []byte(backup), nil
		}

		msg.Body = string(raw)
		return msg, err
	}

	return msg, errors.New("invalid body")
}

func (e *Email) handleMessage(message Message, conds []condition) {
	for _, cond := range conds {
		ts := message.Date.Unix()
		if cond.start.Unix() < ts && ts < cond.end.Unix() &&
			match(message.From, cond.rfrom) &&
			match(message.Body, cond.rbody) &&
			match(message.Subject, cond.rsubj) {
			switch cond.operate.Type {
			case OPDEL:
				e.delete(message)
			case OPREAD:
				e.seen(message)
			case OPMOVE:
				e.move(message, cond)
			}
		}
	}
}

func (e *Email) seen(message Message) {
	fmt.Printf("tag mail to READ from [%v].\n", message.From)
	cmd := Tag{Uid: message.Origin.Uid, Value: "\\Seen"}
	e.client.Execute(cmd, nil)
}

func (e *Email) delete(message Message) {
	fmt.Printf("delete mail from [%v] [%v].\n", message.From, message.Subject)
	set := new(imap.SeqSet)
	set.AddNum(message.Origin.Uid)
	err := e.client.UidStore(set, "+FLAGS", `(\Deleted)`, make(chan *imap.Message))
	fmt.Println(err)
	ch := make(chan uint32, 1)
	ch <- message.Origin.Uid
	go func() {
		e.client.Expunge(ch)
	}()
}

func (e *Email) move(message Message, cond condition) {
	fmt.Printf("move mails from [%v]\n", message.From)
	set := new(imap.SeqSet)
	set.AddNum(message.Origin.Uid)
	cmd := commands.Copy{
		SeqSet:  set,
		Mailbox: cond.operate.MailBox,
	}
	e.client.Execute(&cmd, nil)
	e.delete(message)
}

type Tag struct {
	Uid   uint32
	Value string
}

func (e Tag) Command() *imap.Command {
	set := new(imap.SeqSet)
	set.AddNum(e.Uid)
	store := &commands.Store{
		SeqSet: set,
		Item:   imap.AddFlags,
		Value:  e.Value,
	}
	return store.Command()
}

func main() {
	cfg := `[{
        "name": "github",
        "since": "2021-01-10T00:00:00Z",
        "before": "2022-04-18T00:00:00Z",
        "from": ["github.com"],
        "body": ["jobs have failed", "refs/heads", "All jobs were cancelled",
                 "No jobs were run", "Successfully deployed",
                 "Deployment failed", "A personal access token"],
        "op": {
            "type": "delete"
        }
    }, {
        "name": "google",
        "since": "2021-01-10T00:00:00Z",
        "before": "2022-04-18T00:00:00Z",
        "from": ["no-reply@accounts.google.com"],
        "subject": ["安全提醒", "帐号的安全性"],
        "op": {
           "type": "delete"
        }
    }, {
        "name": "jd",
        "since": "2021-01-10T00:00:00Z",
        "before": "2022-04-18T00:00:00Z",
        "from": ["newsletter1@jd.com"],
        "op": {
            "type": "delete"
        }
    }]`

	var configs []config
	json.Unmarshal([]byte(cfg), &configs)
	fmt.Println("len", len(configs))

	e := Email{
		Username: "",
		Password: "",
	}
	err := e.Login()
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	e.Handle(configs)
}
